src/sctools/fastq.py

```diff
@@ -0,0 +1,197 @@
+from collections import namedtuple
+from . import reader
+
+
+class Record:
+    """Abstract Fastq record object
+
+    Defines properties for accessing fastq record information:
+    :property name: name field
+    :property sequence: sequence field
+    :property name2: second name field
+    :property quality: quality field
+
+    Utilities for parsing read attributes:
+    :property average_quality: return the mean quality of FastqRecord
+    """
+    __slots__ = ['_data']
+
+    def __init__(self, record):
+        """
+        :param [str|bytes] record: list of four strings or bytes objects defining a
+          single fastq record
+        """
+        self._data = list(record)
+
+    @property
+    def name(self):
+        return self._data[0]
+
+    @name.setter
+    def name(self, value):
+        if not isinstance(value, (bytes, str)):
+            raise TypeError('name must be str or bytes')
+        else:
+            self._data[0] = value
+
+    @property
+    def sequence(self):
+        return self._data[1]
+
+    @sequence.setter
+    def sequence(self, value):
+        if not isinstance(value, (bytes, str)):
+            raise TypeError('sequence must be str or bytes')
+        else:
+            self._data[1] = value
+
+    @property
+    def name2(self):
+        return self._data[2]
+
+    @name2.setter
+    def name2(self, value):
+        if not isinstance(value, (bytes, str)):
```

Comment:
Good to have these constraints. In addition to restricting types, are there other constraints that would be worth adding? Do you frequently encounter fields with mangled format that you would want to reject for example?

Reply:
Yea; technically they are supposed to start with '@'. I can add this in.

Reply:
That said, I want to be careful to balance validation with speed; all quality strings should contain only ASCII letters and the name field has a specific structure defined by the sequencer that generates the fastq file. However, these are expensive to validate and might be better served by external validation tools (or expanded later, if we find we want to apply tools that require the structures of those fields to be intact/correct"